tests/test_cmwavex.py

```diff
@@ -0,0 +1,177 @@
+            BETA      42.2967523367957
+            PMLAMBDA           -0.0175
+            PMBETA             -0.3971
+            PX                  0.1515
+            POSEPOCH        55321.0000
+            F0    641.9282333345536244  1  0.0000000000000132
+            F1     -4.330899370129D-14  1  2.149749089617D-22
+            PEPOCH        55321.000000
+            DM               71.016633
+            CM                     0.1
+            TNCHROMIDX               4
+            UNITS                  TDB
+            CMWXFREQ_0001         0.01
+            CMWXCOS_0001             0
+            CMWXCOS_0002             0
+        """
+        get_model(StringIO(par))
+
+    with pytest.raises(ValueError):
+        par = """
+            PSR              B1937+21
+            LAMBDA   301.9732445337270
+            BETA      42.2967523367957
+            PMLAMBDA           -0.0175
+            PMBETA             -0.3971
+            PX                  0.1515
+            POSEPOCH        55321.0000
+            F0    641.9282333345536244  1  0.0000000000000132
+            F1     -4.330899370129D-14  1  2.149749089617D-22
+            PEPOCH        55321.000000
+            DM               71.016633
+            CM                     0.1
+            TNCHROMIDX               4
+            UNITS                  TDB
+            CMWXFREQ_0001         0.00
+            CMWXCOS_0001             0
+        """
+        get_model(StringIO(par))
+
+
+def test_add_cmwavex():
+    par = """
+        PSR              B1937+21
+        LAMBDA   301.9732445337270
+        BETA      42.2967523367957
+        PMLAMBDA           -0.0175
+        PMBETA             -0.3971
+        PX                  0.1515
+        POSEPOCH        55321.0000
+        F0    641.9282333345536244  1  0.0000000000000132
+        F1     -4.330899370129D-14  1  2.149749089617D-22
+        PEPOCH        55321.000000
+        DM               71.016633
+        CM                     0.1
+        TNCHROMIDX               4
+        UNITS                  TDB
+    """
+    m = get_model(StringIO(par))
+    idxs = cmwavex_setup(m, 3600, n_freqs=5)
+
+    with pytest.raises(ValueError):
+        m.components["CMWaveX"].add_cmwavex_component(1, index=5, cmwxsin=0, cmwxcos=0)
+
+    m.components["CMWaveX"].add_cmwavex_component(1, index=6, cmwxsin=0, cmwxcos=0)
+    assert m.num_cmwavex_freqs == len(idxs) + 1
+
+    m.components["CMWaveX"].add_cmwavex_component(
+        1 / u.day, index=7, cmwxsin=0 * cmu, cmwxcos=0 * cmu
+    )
+    assert m.num_cmwavex_freqs == len(idxs) + 2
+
+    m.components["CMWaveX"].add_cmwavex_component(2 / u.day)
+    assert m.num_cmwavex_freqs == len(idxs) + 3
+
+    m.components["CMWaveX"].add_cmwavex_components(
+        np.array([3]) / u.day,
+        cmwxsins=np.array([0]) * cmu,
+        cmwxcoses=np.array([0]) * cmu,
+    )
+    assert m.num_cmwavex_freqs == len(idxs) + 4
+
+    with pytest.raises(ValueError):
```

Comment:
I think all of these statements that raise exceptions should be in separate tests?  So the main test should have statements that succeed, and each failure/exception should be separate?

Reply:
Moved to separate tests